### PR TITLE
Added CopyToAsync(PipeWriter)

### LINKED ISSUE #35513: API Proposal: Add CopyToAsync methods to write from PipeReader to PipeWriter
https://github.com/dotnet/corefx/issues/27246 adds methods for copying from the PipeReader to a Stream and from a Stream to a PipeWriter but we haven't yet added APIs to copy from a PipeReader to a PipeWriter:

```C#
public abstract class PipeReader
{
      public virtual Task CopyToAsync(PipeWriter writer, CancellationToken cancellationToken = default);
}
```

### PR DESCRIPTION
**Pull request by davidfowl**
- Reuse existing CopyToAsync routine to implement the PipeWriter overload.
- Added tests for PipeWriter overload

Fixes #35513

### CODE REVIEW

File: src/System.IO.Pipelines/src/System/IO/Pipelines/PipeReader.cs [hunk 1]
```diff
@@ -89,7 +89,38 @@ public static PipeReader Create(Stream stream, StreamPipeReaderOptions readerOpt
...
+            {
+                FlushResult result = await destination.WriteAsync(memory, cancellationToken).ConfigureAwait(false);
+
+                if (result.IsCanceled)
+                {
+                    ThrowHelper.ThrowOperationCanceledException_FlushCanceled();
+                }
+            }
+
+            return CopyToAsyncCore(destination, WriteAsync, cancellationToken);
```
Comment (stephentoub) [inline src/System.IO.Pipelines/src/System/IO/Pipelines/PipeReader.cs]: This is going to allocate a delegate for WriteAsync on each call; presumably that's not desirable.  In which case you'd be better off not using a local function and instead just using a lambda:
```C#
return CopyToAsyncCore(destination, (destination, memory, cancellationToken) =>
{
    FlushResult result = await destination.WriteAsync(memory, cancellationToken).ConfigureAwait(false);
    if (result.IsCanceled)
    {
        ThrowHelper.ThrowOperationCanceledException_FlushCanceled();
    }
}, cancellationToken);
```

File: src/System.IO.Pipelines/src/System/IO/Pipelines/PipeReader.cs [hunk 2]
```diff
@@ -104,10 +135,15 @@ public virtual Task CopyToAsync(Stream destination, CancellationToken cancellati
                 return Task.FromCanceled(cancellationToken);
             }
 
-            return CopyToAsyncCore(destination, cancellationToken);
+            static ValueTask WriteAsync(Stream destination, ReadOnlyMemory<byte> memory, CancellationToken cancellationToken)
+            {
+                return destination.WriteAsync(memory, cancellationToken);
+            }
+
+            return CopyToAsyncCore(destination, WriteAsync, cancellationToken);
```
Comment (stephentoub) [inline src/System.IO.Pipelines/src/System/IO/Pipelines/PipeReader.cs]: Ditto:
```C#
return CopyToAsyncCore(
    destination,
    (destination, memory, cancellationToken) => destination.WriteAsync(memory, cancellationToken),
    cancellationToken);
```

File: src/System.IO.Pipelines/tests/PipeReaderCopyToAsyncTests.cs [hunk 3]
```diff
@@ -21,7 +22,7 @@ public class CopyToAsyncTests
         public async Task CopyToAsyncThrowsArgumentNullExceptionForNullDestination()
         {
             var pipe = new Pipe(s_testOptions);
-            var ex = await Assert.ThrowsAsync<ArgumentNullException>(() => pipe.Reader.CopyToAsync(null));
+            var ex = await Assert.ThrowsAsync<ArgumentNullException>(() => pipe.Reader.CopyToAsync((Stream)null));
             Assert.Equal("destination", ex.ParamName);
```
Comment (stephentoub) [inline src/System.IO.Pipelines/tests/PipeReaderCopyToAsyncTests.cs]:
```C#
await AssertExtensions.ThrowsAsync<ArgumentNullException>("destination", () => pipe.Reader.CopyToAsync((Stream)null));
```

File: src/System.IO.Pipelines/tests/PipeReaderCopyToAsyncTests.cs [hunk 4]
```diff
@@ -164,6 +194,35 @@ public async Task CancelingViaCancellationTokenThrowsOperationCancelledException
             await Assert.ThrowsAsync<OperationCanceledException>(() => task);
         }
 
+        [Fact]
+        public async Task CancelingPipeWriterViaCancellationTokenThrowsOperationCancelledException()
```
Comment (stephentoub) [inline src/System.IO.Pipelines/tests/PipeReaderCopyToAsyncTests.cs]: Nit: Cancelled => Canceled

File: src/System.IO.Pipelines/src/System/IO/Pipelines/PipeReader.cs (hunk 1 above)
Comment (davidfowl) [reply src/System.IO.Pipelines/src/System/IO/Pipelines/PipeReader.cs]: We never fixed that!?

Comment (stephentoub) [reply src/System.IO.Pipelines/src/System/IO/Pipelines/PipeReader.cs]: https://github.com/dotnet/roslyn/issues/5835

**Pull request merged**